File: Year 1/Academic/Haskell/lab assignments/CatAdmin.hs
--
-- TEST HARNESS FOR THE HASKELL DATABASE EXERCISE
--

module CatAdmin where

import "Database.hs"



--
-- the user-level function
--

admin :: IO ()
admin = interact
        (\input_stream -> intro ++ prompt ++ parse_stream input_stream Empty)
        -- the database is initialized explicitly to the constant Empty,
        -- rather than calling create, so that the menu can appear initially
        -- even for those who have not yet written any of the functions.



--
-- auxiliary support functions
--

intro :: [Char]
intro = "\n"++"           Cattery Administration Tool in Haskell \n\n"  ++
              "                 Version 1.1       \n\n"  ++
              "          TYPE help  for help menu.\n\n\n"


parse_stream :: [Char] -> Database -> [Char]
parse_stream [] d      = "\n"++"cat admin tool logout"
parse_stream input d   = commandline ++ "\n" ++ message ++
				case continue of
				  True  -> prompt ++ parse_stream remains newd
				  False -> []
                         where
                         commandline               = before '\n' input
                         remains                   = after '\n' input
                         command                   = parse_line commandline
                         (continue, message, newd) = processcommand command d


parse_line :: [Char] -> [[Char]]
parse_line [] = []
parse_line (h:t) | isalphanum h = capitalise firstword : parse_line otherwords
                 | otherwise    =  parse_line skip
                   where
                   firstword  = takeWhile isalphanum (h:t)
                   otherwords = dropWhile isalphanum (h:t)
                   skip       = dropWhile notalphanum (h:t)


processcommand :: [[Char]] -> Database -> (Bool, [Char], Database)
processcommand []                d = (True, [], d)
processcommand ("ADD":n:s:t:[])  d | isnum t
					= (True, [], (add n s tval d))
                                   | otherwise
					= (True,
					   "Usage: add name sort tins\n", d)
                                     where tval = numval t
processcommand ("ADD":args)      d = (True, "Usage: add name sort tins\n", d)
processcommand ("DELETE":n:s:[])   d = (True, [], delete (n,s) d)
processcommand ("DELETE":args)   d = (True, "Usage: delete name sort\n", d)
processcommand ("CREATE":args)   d = (True, [], create)
processcommand ("PUT":args) d = (True, (putdatabase d)++"\n", d)
processcommand ("LOOKUP":n:s:[])  d = (True, (do_lookup_cat (n,s) d)++"\n", d)
processcommand ("LOOKUP":args)  d = (True, "Usage: lookup name sort\n", d)
processcommand ("COUNT":args)    d = (True, (do_count d)++"\n", d)
processcommand ("HELP":args)     d = (True, help_menu, d)
processcommand ("QUIT":args)     d = (False, "cat admin tool logout", d)
processcommand (unknown:args)    d = (True, "unrecognised command"++"\n", d)


help_menu :: [Char]
help_menu = "\nHelp Menu\n"++
            "            add name sort tins    - to add a cat\n"++
            "            delete name sort      - to delete a cat\n"++
            "            create                - to restart with a fresh empty database\n"++
                                -- this allows testing of create, which would
                                -- otherwise never be used anywhere!
            "            put                   - to show the database\n"++
            "            lookup name sort      - to look up a cat by name\n"++
            "            count                 - to count cats, sorts, and tins needed\n"++
            "            help                  - to see this menu.\n"++
            "            quit                  - to exit\n\n"



--
-- database-handling code
--

do_lookup_cat :: ([Char],[Char]) -> Database -> [Char]
do_lookup_cat (n,s) d  | t == -1   = n ++ " is not in the database"
                       | otherwise = n ++ " the " ++ s ++ " needs "
                                       ++ show t ++ " tins of cat food weekly"
                         where t = lookup_cat (n,s) d

do_count :: Database -> [Char]
do_count d = "No. of cats = " ++ show n ++ ", number of sorts  = " ++ show s ++
                ", total number of tins needed = " ++ show t
            where (n,s,t) = count d



--
-- low-level stuff.
--

capitalise :: [Char] -> [Char]
capitalise [] = []
capitalise (h:t) = (cap h):(capitalise t)

cap :: Char -> Char
cap ch | islowercase ch = chr ((ord ch) + offset)
       | otherwise      = ch
         where offset = (ord 'A') - (ord 'a')

islowercase :: Char -> Bool
islowercase ch = ((ch >= 'a') && (ch <= 'z'))

isdigit :: Char -> Bool
isdigit ch = ((ch >= '0') && (ch <= '9'))

isnum :: [Char] -> Bool
isnum ('-':t) = isnum t
isnum n = isanumber n

isanumber :: [Char] -> Bool
isanumber [] = False
isanumber (h:[]) = isdigit h
isanumber (h:t)  = (isdigit h) && (isanumber t)

numval :: [Char] -> Int
numval s = numval_aux 0 s
	   where
	   numval_aux acc []    = acc
	   numval_aux acc (h:t) = numval_aux (10*acc+(ord h - ord '0')) t

before :: Char -> [Char] -> [Char]
before ch l  = takeWhile (/= ch) l

isalphanum :: Char -> Bool
isalphanum '.' = True
isalphanum '-' = True
isalphanum ch = (isletter ch) || (isdigit ch)

notalphanum :: Char -> Bool
notalphanum c = not (isalphanum c)

after :: Char -> [Char] -> [Char]
after ch [] = []
after ch l  | drop /= [] = tail drop
            | otherwise  = []
              where drop = dropWhile (/= ch) l

isletter :: Char -> Bool
isletter ch = ((ch >= 'a') && (ch <= 'z')) || ((ch >= 'A') && (ch <= 'Z'))

prompt :: [Char]
prompt = "admin> "

File: Year 1/Academic/Java/Java/DATABASE/CatAdmin.java
/*
 *      CatAdmin.java:
 *      Test harness for implementation class Database.java,
 *	which implements the interface class (DatabaseInterface.java)
 *	offering access procedures for a database of cats
 *	in a cattery and their dietary requirements.
 */

import database.*;

public class CatAdmin
{


public static void main(String[] args)
{
	boolean		finished	= false;
	String		command;
	String		name;
	int		tins;
	Database	db		= new Database();

	System.out.println("Cattery Administration Tool in Java");
	printMenu();

	while (! finished) {
		System.out.print("Select a menu option: ");
		command = kenya.io.InputReader.readString();

		if (command.equals("add")) {
			name = kenya.io.InputReader.readString();
			tins = kenya.io.InputReader.readInt();
			db.add(name, tins);

		} else if (command.equals("lookup")) {
			name = kenya.io.InputReader.readString();
			tins = db.lookup(name);
			System.out.println("NAME\tNO. OF TINS");
			if (tins != -1) {
				System.out.println(name + "\t" + tins);
			} else {
				System.out.println(name + "\t***Not found***");
			}

		} else if (command.equals("count")) {
			System.out.println("NO. OF CATS\tTOTAL NO. OF TINS");
			System.out.println
				(db.countCats() + "\t\t" + db.countTins());

		} else if (command.equals("delete")) {
			name = kenya.io.InputReader.readString();
			db.delete(name);

		} else if (command.equals("print")) {
			System.out.println("NAME\tNO. OF TINS");
			db.printDatabase();

		} else if (command.equals("help")) {
			printMenu();

		} else if (command.equals("quit")) {
			finished = true;

		} else {
			System.out.println
			 ("***Command `" + command + "' not recognized***");
			printMenu();

		}

		System.out.println();
	}

	System.out.println("Thank you for using Cattery Administration Tool");
}


public static void printMenu()
{
	System.out.println("");
        System.out.println("	****** MENU OPTIONS ******");
        System.out.println("");
        System.out.println("	add name tins	- add a cat");
        System.out.println("	delete name	- delete a cat");
        System.out.println("	print		- print the whole database");
        System.out.println("	lookup name	- look up a cat by name");
        System.out.println("	count		- count the cats and tins");
        System.out.println("	help		- show this menu again");
        System.out.println("	quit		- quit!");
        System.out.println("");
}


}
